sdk/storage/azblob/internal/exported/hashing_reader.go

```diff
@@ -0,0 +1,56 @@
+//go:build go1.18
+// +build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+package exported
+
+import (
+	"bytes"
+	"hash/crc64"
+	"io"
+)
+
+const CRC64Polynomial uint64 = 0x9A6C9329AC4BC9B5
+
+var CRC64Table = crc64.MakeTable(CRC64Polynomial)
+
+// ReadWrapper is an io.ReadSeekCloser wrapper to prevent reading twice for SDK-generated hashes.
+type ReadWrapper struct {
+	internal *bytes.Reader
+	crc64    uint64
+}
+
+func (h *ReadWrapper) CRC64Hash() uint64 {
+	return h.crc64
+}
+
+func (h *ReadWrapper) Read(p []byte) (n int, err error) {
+	return h.internal.Read(p)
+}
+
+func (h *ReadWrapper) Seek(offset int64, whence int) (int64, error) {
+	return h.internal.Seek(offset, whence)
+}
+
+func (h *ReadWrapper) Close() error {
+	return nil
+}
+
+func NewReadWrapper(r io.ReadSeekCloser, validationOption TransferValidationType) (*ReadWrapper, error) {
+	buf, err := io.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+
+	var crc uint64
+	if validationOption&TransferValidationTypeCRC64 == TransferValidationTypeCRC64 {
```

Comment:
What happens if this is `false`?  Should it return an error?

Reply:
Absolutely nothing. The hash just doesn't get generated, and the read wrapper works as expected.

The goal of the wrapper was, per Jocelyn's request, to make it so that we didn't have to do a re-read for a hash, and then a re-read for submission.

Should the service adjust to accept HTTP footers, this read wrapper will adjust to generate the hash during the read and then have it be submitted afterwards.